refactor: Use util::Result class for wallet loading

ryanofsky opened the pull request:
<!-- begin based-on -->
**This is based on #25665 + #26022.** The non-base commits are:

- [`b8922344692` Add temporary ResultExtract helper for porting to util::Result](https://github.com/bitcoin/bitcoin/pull/25722/commits/b8922344692ec94a6289c1b38ae36c9605e6110c)
- [`c9445d447d5` refactor: Use util::Result class in wallet/sqlite](https://github.com/bitcoin/bitcoin/pull/25722/commits/c9445d447d5df3c846c63534785e1df19853cedb)
- [`fe479a116d8` refactor: Use util::Result class in wallet/migrate](https://github.com/bitcoin/bitcoin/pull/25722/commits/fe479a116d8a49efe86fc3071710958b91ac3b8b)
- [`f09f74b8336` refactor: Use util::Result class in wallet::MakeDatabase](https://github.com/bitcoin/bitcoin/pull/25722/commits/f09f74b8336ee415a668a50dc0c122b294c649a0)
- [`ac3a5145e53` refactor: Use util::Result class in wallet/dump](https://github.com/bitcoin/bitcoin/pull/25722/commits/ac3a5145e53547208152781a310f88f0b2b5f781)
- [`e8da1aa2096` refactor: Use util::Result class in wallet/wallettool](https://github.com/bitcoin/bitcoin/pull/25722/commits/e8da1aa20964716d7ae93124440c40f5b23a95b7)
- [`711b7b6b7ae` refactor: Use util::Result class in wallet/wallet](https://github.com/bitcoin/bitcoin/pull/25722/commits/711b7b6b7ae65559280a52e228f8c3a832e1f26b)
- [`610ad9dece4` refactor: Use util::Result class in wallet/load](https://github.com/bitcoin/bitcoin/pull/25722/commits/610ad9dece42e2c4ec3e2eb0fded44856f6a65b7)
- [`515737cc9b4` refactor: Use util::Result class in wallet/rpc](https://github.com/bitcoin/bitcoin/pull/25722/commits/515737cc9b48ee076790b487b4291c8b845a8d9f)
- [`e6c4a98d498` refactor: Use util::Result class in wallet/interfaces](https://github.com/bitcoin/bitcoin/pull/25722/commits/e6c4a98d498e87a47f722ac35e222d2e28255a1a)
- [`0afedba86e7` refactor: Use util::Result class in wallet/test](https://github.com/bitcoin/bitcoin/pull/25722/commits/0afedba86e7af48eeb70d8b8e11a7abd8b7bfbdb)
- [`082580fa2c9` Drop temporary ResultExtract helper for porting to util::Result](https://github.com/bitcoin/bitcoin/pull/25722/commits/082580fa2c994c915e01823a3181039203d95ccb)
- [`2bbdb750757` scripted-diff: replace wallet DatabaseStatus with DatabaseError](https://github.com/bitcoin/bitcoin/pull/25722/commits/2bbdb750757d5c079732f632a977bf9680e8a59e)<!-- end -->

---

Wallet loading functions up and down the stack have lots of error and warning parameters, and return error information in different ways. This PR makes them uniformly return `util::Result`, without changing behavior.

ryanofsky committed:
refactor: Add util::Result::Update() method

Add util::Result Update method to make it possible to change the value of an
existing Result object. This will be useful for functions that can return
multiple error and warning messages, and want to be able to change the result
value while keeping existing messages.

ryanofsky committed:
refactor: Add util::Result failure values

Add util::Result support for returning more error information. For better error
handling, adds the ability to return a value on failure, not just a value on
success. This is a key missing feature that made the result class not useful
for functions like LoadChainstate() which produce different errors that need to
be handled differently.

This change also makes some more minor improvements:

- Smaller type size. On 64-bit platforms, util::Result<int> is 16 bytes, and
  util::Result<void> is 8 bytes. Previously util::Result<int> and
  util::Result<void> were 72 bytes.

- More documentation and tests.

ryanofsky committed:
refactor: Use util::Result class in wallet/test

ryanofsky committed:
refactor: Use util::Result class in wallet/load

ryanofsky committed:
refactor: Use util::Result class in wallet/wallet

ryanofsky committed:
refactor: Use util::Result class in wallet/dump

ryanofsky committed:
refactor: Use util::Result class in wallet/sqlite

ryanofsky committed:
refactor: Use util::Result class in wallet/interfaces

ryanofsky committed:
refactor: Use util::Result class in wallet/wallettool

ryanofsky committed:
refactor: Use util::Result class in wallet/migrate

ryanofsky commented:
Rebased a09c21c05a9adf0ca5a4b6a15021b45c6c2e0b2a -> e7457d0acefeccd636c134d0c10567ffaade2fc7 ([`pr/bresult-load.2`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.2) -> [`pr/bresult-load.3`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.2-rebase..pr/bresult-load.3)) due to conflict with #25616

ryanofsky committed:
Use ResultPtr<unique_ptr> instead of Result<unique_ptr>

martinus commented:
After rebasing I get a compile error in `ChooseSelectionResult`, adding a bunch of `std::move` helps

martinus commented on src/wallet/spend.cpp:
```diff
@@ -715,7 +715,7 @@ util::Result<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, Coin
-                return util::Result<SelectionResult>({_("Unconfirmed UTXOs are available, but spending them creates a chain of transactions that will be rejected by the mempool")});
+                return util::Error{_("Unconfirmed UTXOs are available, but spending them creates a chain of transactions that will be rejected by the mempool")};
             }
             return util::Result<SelectionResult>(util::Error()); // General "Insufficient Funds"
```
could be just `return util::Error();`

martinus reviewed (changes requested):
Code review  2a9db07, apart from the compile error when rebasing I found mostly nits. This makes `util::Result` quite a bit more powerful, and it's still simple to use. The code itself of util::Result is a bit hard to understand though, but I don't have a better way either.

TheCharlatan commented on src/util/result.h:
```diff
@@ -31,54 +140,118 @@ struct Error {
+//! applications that do need more complicated error-handling behavior can
+//! override the default `F = void` failure type and get failure values by
+//! calling result.GetFailure().
+template <typename T, typename F = void>
```
I guess the same goes for something like `Result<std::optional<int>, int>`, or really any type implementing `operator bool()`? How did you handle these cases in your code? Sometimes the non-presence of a result is not really an error.

ryanofsky committed:
test: add static test for util::Result memory usage

Suggested by Martin Leitner-Ankerl <martin.ankerl@gmail.com>
https://github.com/bitcoin/bitcoin/pull/25722#discussion_r1174298529

Co-authored-by: Martin Leitner-Ankerl <martin.ankerl@gmail.com>

ryanofsky reviewed (commented):
Rebased f7d4451b98014176c083adc02bdf59e6dd2b9355 -> 17891a95ab5873aa978a5bf5ed8985ee1513e728 ([`pr/bresult-load.16`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.16) -> [`pr/bresult-load.17`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.16-rebase..pr/bresult-load.17) making most of the suggested changes. I'm working on some more changes that should use `ResultPtr` from #26022 to simplify some of the code that has changed and made it safer. It will take more work to make `ResultPtr` compatible, and relatedly change the Result class to discourage using it with confusing types that are convertible to bool.

ryanofsky commented:
Rebased 1b2a5f12b42543d13a4bcafb9585e3d1df488914 -> 6700299bcb6f560e9e4caf34254d7d86d4e78833 ([`pr/bresult-load.19`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.19) -> [`pr/bresult-load.20`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.20), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.19-rebase..pr/bresult-load.20)) on top of #26022 using #26022 to simplify some changes in this PR

ryanofsky committed:
Drop temporary ResultExtract helper for porting to util::Result

ryanofsky committed:
Add temporary ResultExtract helper for porting to util::Result

ryanofsky committed:
scripted-diff: replace wallet DatabaseStatus with DatabaseError

-BEGIN VERIFY SCRIPT-
git grep -l DatabaseStatus src | xargs sed -i s/DatabaseStatus/DatabaseError/g
sed -i '/^    SUCCESS,$/d' src/wallet/db.h
-END VERIFY SCRIPT-

ryanofsky commented:
Updated 6700299bcb6f560e9e4caf34254d7d86d4e78833 -> 2385cb37dbbe00004f8bffe1d1d784778d89a21e ([`pr/bresult-load.20`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.20) -> [`pr/bresult-load.21`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.21), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.20..pr/bresult-load.21)) splitting this up from one big to commit to multiple commits so it is reasonable to review.

However, looking over the commits, this still belongs in draft state, because since the time this PR was opened, a lot of improvements were made to the util::Result class in base PR #25665 based on review feedback, so there are a lot of updates that can be made in this PR to make it nicer.

ryanofsky committed:
refactor: Add util::Result multiple error and warning messages

Add util::Result support for returning warning messages and multiple errors,
not just a single error string. This provides a way for functions to report
errors and warnings in a standard way, and simplifies interfaces.

The functionality is unit tested here, and put to use in followup PR
https://github.com/bitcoin/bitcoin/pull/25722

ryanofsky committed:
refactor: Use util::Result class in wallet/rpc

ryanofsky commented:
Rebased 2385cb37dbbe00004f8bffe1d1d784778d89a21e -> 987b28bbc991868ed305dbfc8f0846dd8b0b893c ([`pr/bresult-load.21`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.21) -> [`pr/bresult-load.22`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.22), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.21-rebase..pr/bresult-load.22)) to use newer result API in #25722.
Updated 987b28bbc991868ed305dbfc8f0846dd8b0b893c -> 548bf508e26aae39c732b3e61861d5c29669df06 ([`pr/bresult-load.22`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.22) -> [`pr/bresult-load.23`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.23), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.22..pr/bresult-load.23)) to fix a bug in wallet tool and an accidental change to error message ordering
Updated 548bf508e26aae39c732b3e61861d5c29669df06 -> ae0b9b8fcece154bae7f4383990e9d10f6cc0d2b ([`pr/bresult-load.23`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.23) -> [`pr/bresult-load.24`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.24), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.23..pr/bresult-load.24)) to fix more unintended changes to error message order.
Updated ae0b9b8fcece154bae7f4383990e9d10f6cc0d2b -> 5beaefaa806e54327d5995d86fd2d56d4687d575 ([`pr/bresult-load.24`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.24) -> [`pr/bresult-load.25`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.25), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.24..pr/bresult-load.25)) to fix clang-tidy errors in https://cirrus-ci.com/task/6679456805289984 and ResultExtract bug in intermediate commit https://github.com/bitcoin/bitcoin/actions/runs/8931046064/job/24532379963?pr=25722

ryanofsky committed:
refactor: Use util::Result class in wallet::MakeDatabase

ryanofsky committed:
wallet: fix clang-tidy warning performance-no-automatic-move

Previous commit causes the warning to be triggered, but the suboptimal return
from const statement was added earlier in 517e204bacd9d.

Warning was causing CI failure https://cirrus-ci.com/task/6159914970644480

ryanofsky committed:
refactor: Use util::Result class in LoadChainstate and VerifyLoadedChainstate

ryanofsky commented:
Rebased 2b0e70295ed6ff4be91cc5775ddaa9351a1ed430 -> caed6aec4241a2e5d849c9446725c889348e1b7e ([`pr/bresult-load.30`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.30) -> [`pr/bresult-load.31`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.31), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.30-rebase..pr/bresult-load.31)) due to various conflicts
Updated caed6aec4241a2e5d849c9446725c889348e1b7e -> 830570fe684e05c170def209633c682363d3b830 ([`pr/bresult-load.31`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.31) -> [`pr/bresult-load.32`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.32), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.31..pr/bresult-load.32)) to fix DoMigration error handling bug https://github.com/bitcoin/bitcoin/actions/runs/11665336901/job/32477794828?pr=25722
Rebased 830570fe684e05c170def209633c682363d3b830 -> e2376f9847c8cd07bb5f2c82f8b0055a72086b66 ([`pr/bresult-load.32`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.32) -> [`pr/bresult-load.33`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.33), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.32-rebase..pr/bresult-load.33)) due to various conflicts
Rebased e2376f9847c8cd07bb5f2c82f8b0055a72086b66 -> 0284396a4229025a6201d7486484b17c00f3eb8c ([`pr/bresult-load.33`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.33) -> [`pr/bresult-load.34`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.34), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.33-rebase..pr/bresult-load.34)) due to conflicts with #31241, #31451, and other PRs
Rebased 0284396a4229025a6201d7486484b17c00f3eb8c -> 88dfdac4bc23c87f79c55ed6b2c255f29137768e ([`pr/bresult-load.34`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.34) -> [`pr/bresult-load.35`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.35), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.34-rebase..pr/bresult-load.35)) due to conflicts with #31961 and #31519
<!-- begin push-36 -->
Rebased 88dfdac4bc23c87f79c55ed6b2c255f29137768e -> 96f66136f02833f00520eb18e32d0a2848cb4c2a ([`pr/bresult-load.35`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.35) -> [`pr/bresult-load.36`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.36), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.35-rebase..pr/bresult-load.36))<!-- end -->
<!-- begin push-37 -->
Rebased 96f66136f02833f00520eb18e32d0a2848cb4c2a -> fdff2b4550e8b2b64e25d413298283fa1b516088 ([`pr/bresult-load.36`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.36) -> [`pr/bresult-load.37`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.37), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.36-rebase..pr/bresult-load.37))<!-- end -->
<!-- begin push-38 -->
Rebased fdff2b4550e8b2b64e25d413298283fa1b516088 -> 46d01725ed1434fba8d974f57f6893e2cd1facee ([`pr/bresult-load.37`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.37) -> [`pr/bresult-load.38`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.38), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.37-rebase..pr/bresult-load.38))<!-- end --> due to conflict with #30595
<!-- begin push-39 -->
Rebased 46d01725ed1434fba8d974f57f6893e2cd1facee -> 2dc6d076e76f9c778dbced9adfd0ef64aea732b1 ([`pr/bresult-load.38`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.38) -> [`pr/bresult-load.39`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.39), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.38-rebase..pr/bresult-load.39))<!-- end --> on top of updated base PR to fix CI errors
<!-- begin push-40 -->
Rebased 2dc6d076e76f9c778dbced9adfd0ef64aea732b1 -> d3e677f59d6369dc295edb7b5f112b83f99ec3c9 ([`pr/bresult-load.39`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.39) -> [`pr/bresult-load.40`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.40), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.39-rebase..pr/bresult-load.40))<!-- end -->
<!-- begin push-41 -->
Rebased d3e677f59d6369dc295edb7b5f112b83f99ec3c9 -> 2c7996d7c398b0927008a1903aad786d81c3e949 ([`pr/bresult-load.40`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.40) -> [`pr/bresult-load.41`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.41), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.40-rebase..pr/bresult-load.41))<!-- end -->
<!-- begin push-42 -->
Rebased 2c7996d7c398b0927008a1903aad786d81c3e949 -> 06c3f0abda7bf258993c7c1742502d511574a50e ([`pr/bresult-load.41`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.41) -> [`pr/bresult-load.42`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.42), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.41-rebase..pr/bresult-load.42))<!-- end --> on top of #26022 pr/bresult-ptr.17
<!-- begin push-43 -->
Updated 06c3f0abda7bf258993c7c1742502d511574a50e -> 45da1d9a13b26d5030354c7d5d874fd59cf35d84 ([`pr/bresult-load.42`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.42) -> [`pr/bresult-load.43`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.43), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.42..pr/bresult-load.43))<!-- end --> to fix CI errors https://github.com/bitcoin/bitcoin/actions/runs/21101803730/job/60687048929?pr=25722
<!-- begin push-44 -->
Rebased 45da1d9a13b26d5030354c7d5d874fd59cf35d84 -> 23df7666d1cfae87609b0054e7184bfad71467dd ([`pr/bresult-load.43`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.43) -> [`pr/bresult-load.44`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.44), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.43-rebase..pr/bresult-load.44))<!-- end --> on top of #25665 pr/bresult2.73 due to conflicts with #31650 and #34269
<!-- begin push-45 -->
Added 1 commits 23df7666d1cfae87609b0054e7184bfad71467dd -> 9382817901067c57b5ff4044d4c7aa09aee25376 ([`pr/bresult-load.44`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.44) -> [`pr/bresult-load.45`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.45), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.44...pr/bresult-load.45))<!-- end --> to fix ci iwyu error https://github.com/bitcoin/bitcoin/actions/runs/21266012856/job/61207857203?pr=25722
<!-- begin push-46 -->
Rebased 9382817901067c57b5ff4044d4c7aa09aee25376 -> e164e1e7a5c89eab4c0ef30a4a772a249a95ce0c ([`pr/bresult-load.45`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.45) -> [`pr/bresult-load.46`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.46), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.45-rebase..pr/bresult-load.46))<!-- end --> due to conflicts with #32636
<!-- begin push-47 -->
Updated e164e1e7a5c89eab4c0ef30a4a772a249a95ce0c -> 8253c7087f2cd7f5d0e3e35f39120e29557be846 ([`pr/bresult-load.46`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.46) -> [`pr/bresult-load.47`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.47), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.46..pr/bresult-load.47))<!-- end --> to fix tidy and iwyu errors https://github.com/bitcoin/bitcoin/actions/runs/21833918575/job/62999419836?pr=25722
<!-- begin push-48 -->
Rebased 8253c7087f2cd7f5d0e3e35f39120e29557be846 -> d75ff7517eb28aa549a0fcfaa93c7fa20f391055 ([`pr/bresult-load.47`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.47) -> [`pr/bresult-load.48`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.48), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.47-rebase..pr/bresult-load.48))<!-- end --> on top of #25665 pr/bresult2.76
<!-- begin push-49 -->
Rebased d75ff7517eb28aa549a0fcfaa93c7fa20f391055 -> 89c866499cfc52496327cef6294b40ac7a3ae512 ([`pr/bresult-load.48`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.48) -> [`pr/bresult-load.49`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.49), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.48-rebase..pr/bresult-load.49))<!-- end --> on top of #25665 pr/bresult2.77
<!-- begin push-50 -->
Rebased 89c866499cfc52496327cef6294b40ac7a3ae512 -> 04c5995eec3ff5d82f097069ebf6f33715e70513 ([`pr/bresult-load.49`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.49) -> [`pr/bresult-load.50`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.50), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.49-rebase..pr/bresult-load.50))<!-- end --> on top of #25665 pr/bresult2.79 and pr/bresult-ptr.20 #26022
<!-- begin push-51 -->
Rebased 04c5995eec3ff5d82f097069ebf6f33715e70513 -> 91c331c76d34ad0a5ef112543af984a832d8a039 ([`pr/bresult-load.50`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.50) -> [`pr/bresult-load.51`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.51), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.50-rebase..pr/bresult-load.51))<!-- end --> on top of #25665 pr/bresult2.79 and #26022 pr/bresult-ptr.20 due to conflict with #28333
<!-- begin push-52 -->
Rebased 91c331c76d34ad0a5ef112543af984a832d8a039 -> a268356d39fbad77c871a006164a9f9e58e39d90 ([`pr/bresult-load.51`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.51) -> [`pr/bresult-load.52`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.52), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.51-rebase..pr/bresult-load.52))<!-- end --> on top of #26022 pr/bresult-ptr.21 due to conflict with #35182 (libevent replacement)
<!-- begin push-53 -->
Rebased a268356d39fbad77c871a006164a9f9e58e39d90 -> 2040c025062f30626fd998ee6f582a771f1913f4 ([`pr/bresult-load.52`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.52) -> [`pr/bresult-load.53`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.53), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.52-rebase..pr/bresult-load.53))<!-- end --> due to conflict with #35266
<!-- begin push-54 -->
Updated 2040c025062f30626fd998ee6f582a771f1913f4 -> 036ce1d818c98bba047fb11b978c99b72de55e04 ([`pr/bresult-load.53`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.53) -> [`pr/bresult-load.54`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.54), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.53..pr/bresult-load.54))<!-- end --> to fix iwyu errors (util/translation.h → util/result.h in bench files) https://github.com/bitcoin/bitcoin/actions/runs/28131393626/job/83308198030
<!-- begin push-55 -->
Rebased 036ce1d818c98bba047fb11b978c99b72de55e04 -> 2bbdb750757d5c079732f632a977bf9680e8a59e ([`pr/bresult-load.54`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.54) -> [`pr/bresult-load.55`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.55), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.54-rebase..pr/bresult-load.55))<!-- end --> on top of #26022 pr/bresult-ptr.22 due to conflicts with #35655
<!-- begin push-56 -->
Rebased 2bbdb750757d5c079732f632a977bf9680e8a59e -> 090e9ba606bf9a217bcd86393424e5c3934e4cf5 ([`pr/bresult-load.55`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.55) -> [`pr/bresult-load.56`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.56), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.55-rebase..pr/bresult-load.56))<!-- end --> on top of #26022 pr/bresult-ptr.23 due to conflicts with bitcoin-core/gui#953 and #35914
<!-- begin push-57 -->
Rebased 090e9ba606bf9a217bcd86393424e5c3934e4cf5 -> 19e31d14092f2a6189f834168d60d373470b410b ([`pr/bresult-load.56`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.56) -> [`pr/bresult-load.57`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-load.57), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-load.56-rebase..pr/bresult-load.57))<!-- end --> on top of #26022 pr/bresult-ptr.23 due to conflicts with #35955

ryanofsky committed:
Add util::ResultPtr class

The util::ResultPtr class is a wrapper for util::Result providing syntax sugar
to make it less awkward to use with returned pointers. Instead of needing to be
deferenced twice with **result or (*result)->member, it only needs to be
dereferenced once with *result or result->member. Also when it is used in a
boolean context, instead of evaluating to true when there is no error and the
pointer is null, it evaluates to false so it is straightforward to determine
whether the result points to something.

This PR is only partial a solution to #26004 because while it makes it easier
to check for null pointer values, it does not make it impossible to return a
null pointer value inside a successful result. It would be possible to enforce
that successful results always contain non-null pointers by using a not_null
type (https://github.com/bitcoin/bitcoin/issues/24423) in combination with
ResultPtr, though.

ryanofsky committed:
ci: Avoid -Wno-error=maybe-uninitialized false positives

Avoid false positive maybe-uninitialized errors in native_previous_releases
CI job.

Fix was suggested by maflcko in
https://github.com/bitcoin/bitcoin/pull/25665#issuecomment-3457573218

CI errors look like:
https://github.com/bitcoin/bitcoin/actions/runs/25426711469/job/74581986567?pr=25665

/home/admin/actions-runner/_work/_temp/src/wallet/test/util.cpp: In function ‘wallet::DescriptorScriptPubKeyMan* wallet::CreateDescriptor(CWallet&, const std::string&, bool)’:
/home/admin/actions-runner/_work/_temp/src/wallet/test/util.cpp:143:10: error: ‘*(const std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>*)((char*)&spkm + offsetof(util::Result<std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>, void, util::Messages>,util::Result<std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>, void, util::Messages>::<unnamed>.util::detail::SuccessHolder<std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>, void, util::Messages>::<unnamed>)).std::reference_wrapper<wallet::DescriptorScriptPubKeyMan>::_M_data’ may be used uninitialized [-Werror=maybe-uninitialized]
  143 |     auto spkm = Assert(keystore.AddWalletDescriptor(w_desc, keys,/*label=*/"", /*internal=*/false));
      |          ^~~~
cc1plus: all warnings being treated as errors

ryanofsky committed:
Merge branch 'pr/bresult-ptr' into pr/bresult-load

ryanofsky committed:
Merge branch 'pr/bresult2' into pr/bresult-ptr
